DistroLauncher/Ubuntu/InitTasks.cpp

```diff
@@ -0,0 +1,427 @@
+void waitForInitTasks(WslApiLoader& api) {
+  _putws(L"Checking for initialization tasks...\n");
+
+  DWORD exitCode_ = -1;
+  if (auto hr = api.WslLaunchInteractive(L"which -s cloud-init", FALSE, &exitCode_); FAILED(hr)) {
+    Helpers::PrintErrorMessage(hr);
+    return;
+  }
+  if (exitCode_ != 0) {
+    _putws(L"CheckInitTasks: this release doesn't support initialization tasks.\n");
+    return;
+  }
+
+  exitCode_ = -1;
+  if (auto hr = api.WslLaunchInteractive(L"cloud-init status --wait", FALSE, &exitCode_);
+      FAILED(hr)) {
+    Helpers::PrintErrorMessage(hr);
+    return;
+  }
+  // 0 for "success" or 2 for "recoverable error" - sometimes we get that while status shows "done".
+  // https://cloudinit.readthedocs.io/en/latest/explanation/failure_states.html#cloud-init-error-codes
+  // If that's the case we should inspect the system to check whether we should skip or proceed with
+  // user creation.
+  if (exitCode_ != 0 && exitCode_ != 2) {
+    wprintf(L"CheckInitTasks: failed with exit code: %u\n", exitCode_);
+  }
+}
+
+// Groups the pieces of information from a single user entry in the passwd database we care about.
+struct UserEntry {
+  std::string name;
+  ULONG uid = -1;
+};
+
+namespace fs = std::filesystem;
+fs::path wslConfPath() {
+  // init-once, lazily.
+  static fs::path etcWslConf =
+      fs::path{L"\\\\wsl.localhost"} / DistributionInfo::Name / "etc\\wsl.conf";
+  return etcWslConf;
+}
+
+void writeUserToWslConf(std::string_view name) {
+  auto oldPath = wslConfPath();
+  auto newPath = wslConfPath();
+  newPath.replace_extension(".new");
+
+  {
+    // Write the new file starting with the [user] section.
+    // ios::binary is needed otherwise '\r' will be added on Windows :(.
+    std::ofstream confNew{newPath, std::ios::binary};
+    confNew << "[user]\ndefault=" << name << '\n';
+
+    // Read existing content.
+    std::ifstream confOld{oldPath};
+    std::string original{std::istreambuf_iterator<char>(confOld), std::istreambuf_iterator<char>()};
+    if (!original.empty()) {
+      // and copy that to the new file.
+      std::copy(original.begin(), original.end(), std::ostreambuf_iterator<char>(confNew));
+    }
+  }
+
+  // Files are closed by now, we may replace the old file with the new one.
+  fs::rename(newPath, oldPath);
+}
+
+bool setDefaultUserViaWslApi(WslApiLoader& api, unsigned long uid) {
+  if (auto hr = api.WslConfigureDistribution(uid, WSL_DISTRIBUTION_FLAGS_DEFAULT); FAILED(hr)) {
+    Helpers::PrintErrorMessage(hr);
+    return false;
+  }
+  return true;
+}
+
+// Collects all users found in the NSS passwd database, sorted by UID.
+std::vector<UserEntry> getAllUsers(WslApiLoader& api);
+
+// Returns the defaultUser set in /etc/wsl.conf or the empty string if none is set.
+std::string defaultUserInWslConf();
+
+// Converts a multi-byte into a wide string.
+std::wstring str2wide(const std::string& str, UINT codePage = CP_THREAD_ACP);
+
+bool enforceDefaultUser(WslApiLoader& api) try {
```

Comment:
it’s probably me, but I didn’t see the case which comes in second which is "there is a key registry, then we consider this was the default set for that distro?" as the order is wsl.conf > registry?

Reply:
Added a comment to clarify that. We check indirectly by calling ´DistributionInfo::QueryUid(L"")` which evaluates the effect, i.e. it'd return the UID set via registry or WSL API.